service/kernel/oneapi/cl_kernels/op_reducer.cl

```diff
@@ -0,0 +1,164 @@
+/* file: op_reduce.cl */
+/*******************************************************************************
+* Copyright 2020 Intel Corporation
+*
+* Licensed under the Apache License, Version 2.0 (the "License");
+* you may not use this file except in compliance with the License.
+* You may obtain a copy of the License at
+*
+*     http://www.apache.org/licenses/LICENSE-2.0
+*
+* Unless required by applicable law or agreed to in writing, software
+* distributed under the License is distributed on an "AS IS" BASIS,
+* WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+* See the License for the specific language governing permissions and
+* limitations under the License.
+*******************************************************************************/
+
+/*
+//++
+//  Implementation of reduction kernels.
+//--
+*/
+
+#ifndef __OP_REDUCER_CL__
+#define __OP_REDUCER_CL__
+
+#define DECLARE_SOURCE(name, src) static const char * name = #src;
+
+DECLARE_SOURCE(
+    op_reduce,
+
+    inline algorithmFPType pow2(const algorithmFPType x) { return x * x; }
+
+    inline algorithmFPType none(const algorithmFPType x) { return x; }
+
+    inline algorithmFPType sum(const algorithmFPType x, const algorithmFPType y) { return x + y; }
+
+    __kernel void reduceSinglepass(uint vectorsAreRows, __global algorithmFPType * vectors, uint nVectors, uint vectorSize,
+                                   __global algorithmFPType * reduces) {
+        const uint local_size = get_local_size(0);
+
+        __local algorithmFPType partial_reduces[LOCAL_BUFFER_SIZE];
+
+        uint globalDim = 1;
+        uint localDim  = nVectors;
+
+        if (vectorsAreRows != 0)
+        {
+            globalDim = vectorSize;
+            localDim  = 1;
+        }
+
+        uint itemId  = get_local_id(0);
+        uint groupId = get_global_id(1);
+
+        algorithmFPType el      = vectors[groupId * globalDim + itemId * localDim];
+        partial_reduces[itemId] = INIT_VALUE;
```

Comment:
Why `INIT_VALUE` cannot be taken as an input argument to the kernel?

Reply:
In principle, this can be done with help  dispatcher before calling the kernel and forwarding this argument to other internal kernels. But it will probably it looking worse in my opinion.